linux/ibus-keyman/src/engine.c

```diff
@@ -678,16 +707,59 @@ process_capslock_action(
   return TRUE;
 }
 
-static gboolean process_end_action(IBusKeymanEngine *keyman) {
-  keyman->firstsurrogate = 0;
-  if (keyman->char_buffer != NULL) {
-    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
-    g_free(keyman->char_buffer);
-    keyman->char_buffer = NULL;
+static void
+commit_text(IBusKeymanEngine *keyman) {
+  g_assert(keyman != NULL);
+  if (keyman->commit_item > keyman->commit_queue) {
+    commit_queue_item *current_item = &keyman->commit_queue[0];
+    keyman->firstsurrogate = 0;
+    if (current_item->char_buffer != NULL) {
+      ibus_keyman_engine_commit_string(keyman, current_item->char_buffer);
+      g_free(current_item->char_buffer);
+    }
+    if (current_item->emitting_keystroke) {
+      ibus_engine_forward_key_event((IBusEngine*)keyman, current_item->keyval, current_item->keycode, current_item->state);
+    }
+    keyman->commit_item--;
+    memmove(keyman->commit_queue, &keyman->commit_queue[1], sizeof(commit_queue_item) * MAX_QUEUE_SIZE - 1);
+    initialize_queue(keyman, MAX_QUEUE_SIZE - 1, 1);
   }
-  if (keyman->emitting_keystroke) {
-    keyman->emitting_keystroke = FALSE;
-    return FALSE;
+}
+
+static gboolean
+process_end_action(IBusKeymanEngine *keyman) {
+  g_assert(keyman != NULL);
+  IBusEngine *engine = (IBusEngine *)keyman;
+  if ((engine->client_capabilities & IBUS_CAP_PREFILTER) &&
+      !(engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT)) {
```

Comment:
Given we use this test in more than one place, can we define a function for it? I see usages on lines 667, 733, 843, and 923 (inverted).